pkg/segment/set.go

```diff
@@ -932,11 +1001,150 @@ func (n *node) rebalanceAfterRemove(gap GapIterator) GapIterator {
 		}
 		p.children[p.nrSegments] = nil
 		p.nrSegments--
+		// update maxGap of left, no need to change p and right
+		// right is already invalid
+		left.updateLocalMaxGap()
 		// This process robs p of one segment, so recurse into rebalancing p.
 		n = p
 	}
 }
 
+// update maxGap bottom up from the calling leaf until no updates
+// this should only be called on leaf node with parameter as 0
```

Comment:
I did find it a bit weird that all the callers pass in a random 0 as the parameter here. I suggest splitting this function into two parts, a recursive one and a non-recursive one. E.g.

// updateMaxGapLeaf updates the maximum gap on this node and all ancestors.
// 
// Precondition: should only be called on a leaf node n.
func (n *node) updateMaxGapLeaf() {
   ..
   n.updateMaxGapInternal(...)
}

// updateMaxGapInternal updates the maximum gap on internal nodes recursively.
//
// Precondition: only called by updateMaxGapLeaf.
func (n *node) updateMaxGapInternal(newMaxGap Key) {
    ...
}

Reply:
I have actually though of this when I wrote this function but I think it's unecessary and made the code more ugly to some extent.
If you think it's very necessary, I can do the modification.